### PR TITLE
ensure grammatical correct alternatives with 'zu' in a verb

### LINKED ISSUE #710: "sicherzustellen" ie. zu
Wir haben noch viel sicherzustellen\.

<p></p>

### PR DESCRIPTION
**Pull request by lsmith77**
fixes https://github.com/witty-works/nlp_api/issues/710

image

### CODE REVIEW

File: app/main.py [hunk 1]
```diff
@@ -2341,7 +2341,34 @@ def align_verb_form(lang, a_token, b_token):
 
         return b_text
     elif lang.lang == "de":
-        ending = a_token.text[len(a_token.lemma_) :]
+        # check if "zu" was stripped from the word in the lemma
+        if "zu" in a_token.text and "zu" not in a_token.lemma_:
+            # find verb at the end
+            if (
+                not b_token.text.startswith("be")
```
Comment (lsmith77) [inline app/main.py]: does it make sense to skip all words with `be` (`bewirken`), `er` (`ermöglichen`) and `ge` (`gewährleisten`) at the start? are there other relevant prefixes?

**Comment by lsmith77** [comment]
tried various splitter libs but they don't seem to be able to split up words like `austauschen`:

* https://github.com/dtuggener/CharSplit
* https://github.com/TimKam/compound-word-splitter
* https://github.com/karlb/wikdict-compound
* https://github.com/mammothb/symspellpy

Didn't try out

* https://github.com/repodiac/german_compound_splitter (CC license and unclear data source license)
*

File: app/main.py (hunk 1 above)
Comment (tje-ww) [reply app/main.py]: Prefixes of verbs that can split with “zu”
ab-
an-
auf-
aus-
bei-
ein-
mit-
nach-
weg-
zu-
her-
nach-


For these prefixes, it depends on the verb, whether you can split it with zu or not:
um-
über-
durch-
unter-
um-

For these cases, maybe a verb list would help? https://www.verblisten.de/listen/verben/anfangsbuchstabe/zusammengesetzt/vollstaendig.html

Comment (tje-ww) [reply app/main.py]: You don’t split verbs with the prefixes ge-, er-, ent-, emp-, be-, hinter-, miss-, ver- and zer-.

Comment (lsmith77) [reply app/main.py]: ok thank you .. but what about `sicherzustellen` .. is this not correct? 
https://de.wiktionary.org/wiki/sicherzustellen

Based on what you said above this should be `zu sicherstellen` which is incorrect.

So should `sicher` be added to the list of `it depends`?

what about `hochstellen` => `hochzustellen` 

I think there is still a gap in the above rules.

Comment (tje-ww) [reply app/main.py]: Good catch, that’s another type of compound verb entirely - adjective + verb.
You can split them with “zu.” The big BUT: when you use the compound verb figuratively you spell it as one word, when you use it in the literal sense, you write separate words. 
Example
sicher(zu)stellen, das jemand in der Praxis ist
Opas alte Kristallkelche sicher (zu) stellen, damit sie nicht umfallen
Not sure, if we can catch that (yet). 

That said, there will be different spelling conventions (alte deutsche Rechtschreibung versus neue deutsche Rechtschreibung; Österreichische und Schweizerische Rechtschreibung) I’ll look that up when I get home (approx. 4:30 pm).

**Review by lnazarenko — APPROVED** [review]

**Pull request merged**